test/mode_permutation_test.py

```diff
@@ -0,0 +1,170 @@
+initial_state = [1, 1, 1, 0, 0]
+
+permutation_matrix[0][2] = 1
+permutation_matrix[1][0] = 1
+permutation_matrix[2][3] = 1
+permutation_matrix[3][4] = 1
+permutation_matrix[4][1] = 1
+
+# Define some constants that will be used through the experiment.
+NUMBER_OF_LOST_PARTICLES = 2
+INITIAL_NUMBER_OF_PARTICLES = initial_state.count(1)
+NUMBER_OF_PARTICLES_LEFT = INITIAL_NUMBER_OF_PARTICLES - NUMBER_OF_LOST_PARTICLES
+NUMBER_OF_MODES = 5
+
+# Define some variables for this scenario, to ease the computations.
+possible_outcomes = [[1, 0, 0, 0, 0], [0, 1, 0, 0, 0], [0, 0, 0, 0, 1]]
+
+
+def main():
+    exact_distribution = calculate_exact_distribution()
+    print(f'Exact distribution: {exact_distribution}')
+
+    approximate_distribution = calculate_approximate_distribution()
+    print(f'Approximate distribution: {approximate_distribution}')
+
+
+def calculate_exact_distribution() -> list:
+    """
+    This method will be used to calculate the exact distribution of lossy boson sampling experiment.
+    The results will be returned as a table of probabilities of obtaining the outcome at i-th index.
+    :return: List of probabilities of outcomes.
+    """
+    outcomes_probabilities = []
+    for outcome in possible_outcomes:
+        outcomes_probabilities.append(calculate_probability_of_outcome(outcome))
+    return outcomes_probabilities
+
+
+def calculate_probability_of_outcome(outcome: list) -> float:
+    """
+    Given input state and losses as described in the constants of this file calculate the probability
+    of obtaining the outcome given as a parameter, after traversing lossy linear-optical channel.
+    :param outcome: An outcome which probability of obtaining will be calculated.
+    :return: Probability of obtaining given outcome in situation presented by by the
+    """
+    outcome_probability = 0  # Initialize with 0 for later debug purposes.
+    outcome_state_in_particle_basis = modes_state_to_particle_state(outcome, NUMBER_OF_PARTICLES_LEFT)
+
+    outcome_probability = calculate_probability_of_outcome_state_for_indistinguishable_photons(outcome_state_in_particle_basis)
+
+    # Different states in particles-basis may give the same outcome state.
+    outcome_probability *= math.factorial(NUMBER_OF_PARTICLES_LEFT)
+    for i in range(NUMBER_OF_MODES):
+        outcome_probability /= math.factorial(outcome[i])
+
+    return outcome_probability
+
+
+def calculate_probability_of_outcome_state_for_indistinguishable_photons(outcome_state_in_particle_basis: list) \
+        -> float:
+    copy_of_outcome_state = list(outcome_state_in_particle_basis[:])
+    outcome_state_in_mode_basis = particle_state_to_modes_state(copy_of_outcome_state, NUMBER_OF_MODES)
+
+    probability_of_outcome = 0
+
+    # Symmetrize the input.
+    lossy_inputs_list = generate_lossy_inputs(initial_state, NUMBER_OF_PARTICLES_LEFT)
+    for lossy_input in lossy_inputs_list:
+        subprobability = abs(calculate_permanent(
```

Comment:
Are the values here complex numbers? Because otherwise, you can skip the `abs` part.

Reply:
Yes, we've got complex numbers here (in general). But thanks for vigilance. ;)